core/stats_handler.go

```diff
@@ -0,0 +1,203 @@
+package core
+
+import (
+	"context"
+	"time"
+
+	tpb "github.com/textileio/go-threads/api/pb"
+	hpb "github.com/textileio/textile/v2/api/hubd/pb"
+	mdb "github.com/textileio/textile/v2/mongodb"
+	"google.golang.org/grpc/stats"
+)
+
+type StatsHandler struct {
+	t *Textile
+}
+
+var _ stats.Handler = (*StatsHandler)(nil)
+
+var statsTimeout = time.Second * 10
+
+// HandleRPC accounts for customer usage across services.
+func (h *StatsHandler) HandleRPC(ctx context.Context, st stats.RPCStats) {
+	switch st := st.(type) {
+	case *stats.OutPayload:
+		if getStats(ctx) == nil {
+			return
+		}
+
+		// Handle payload types.
+		egress := int64(st.WireLength)
+		var reads, writes int64
+		var pl interface{}
+		switch spl := st.Payload.(type) {
+		case *tpb.ReadTransactionReply:
+			pl = spl.Option
+		case *tpb.WriteTransactionReply:
+			pl = spl.Option
+		default:
+			pl = spl
+		}
+		switch pl := pl.(type) {
+
+		// Don't charge for usage when customer is trying to add a payment method, cancel subscription, etc.
+		case *hpb.GetBillingSessionResponse:
+			egress = 0
+
+		// Account for threaddb reads and writes
+		case *tpb.CreateReply:
+			if pl.InstanceIDs != nil {
+				writes = int64(len(pl.InstanceIDs))
+			}
+		case *tpb.WriteTransactionReply_CreateReply:
```

Comment:
Oh interesting, of course this makes sense. But it also makes me realize, we have to be pretty careful with our clients now, so that we're "optimizing" calls on behalf of the user/developer.

Reply:
Yep, for sure.